File: java/ch17-03-decimal/src/main/java/com/thesoftwaregorilla/tdd/money/Sum.java
package com.thesoftwaregorilla.tdd.money;

import java.math.BigDecimal;

class Sum extends Expression {
    final Expression augend;
    final Expression addend;

    Sum(Expression augend, Expression addend) {
        this.augend = augend;
        this.addend = addend;
    }

    @Override
    public Money reduce(Bank bank, String to) {
        return new Money(augend.reduce(bank, to).getAmount().add(addend.reduce(bank, to).getAmount()), to);
    }

    @Override
    public Expression times(BigDecimal multiplier) {
        return new Sum(augend.times(multiplier), addend.times(multiplier));
    }
}

File: cs/ch17-02-plus/Money/Sum.cs
namespace TheSoftwareGorilla.TDD.Money;

public class Sum : Expression
{
    public Expression Augend { get; }
    public Expression Addend { get; }

    public Sum(Expression augend, Expression addend)
    {
        Augend = augend;
        Addend = addend;
    }

    public override Money Reduce(Bank bank, string to)
    {
        int amount = Augend.Reduce(bank, to).Amount + Addend.Reduce(bank, to).Amount;
        return new Money(amount, to);
    }


    public override Expression Times(int multiplier)
    {
        return new Sum(Augend.Times(multiplier), Addend.Times(multiplier));
    }
}